### PR TITLE
[Cleanup] Head Tracker Generalisation Interfaces Cleanup

### PR DESCRIPTION
**Pull request by ExtremelySunnyYK**
Refactored use of `Httypes` interfaces into EVM types alias that uses generic types in `common` folder instead [[1464]](https://smartcontract-it.atlassian.net/browse/BCI-1464)

Note:
- In order to mock a nil return value from the headtracker function, a workaround is necessary. This is because the generic type constraint H enforces a specific type, leading to a panic when attempting to convert a nil value to a *types.Head. The panic message is "interface conversion: interface {} is nil, not *types.Head [recovered]".
- EVM headsaver requires a composite interface due to its interface with DB which not all headsaver needs. For instance if headsaver was in memory, that would not be needed.

Ticket:
https://smartcontract-it.atlassian.net/browse/BCI-1464

### CODE REVIEW

File: core/chains/evm/headtracker/types/types.go [hunk 1]
```diff
@@ -7,50 +7,20 @@ import (
...
 )
 
 // HeadSaver maintains chains persisted in DB. All methods are thread-safe.
 type HeadSaver interface {
-	// Save updates the latest block number, if indeed the latest, and persists
-	// this number in case of reboot.
-	Save(ctx context.Context, head *evmtypes.Head) error
-	// Load loads latest EvmHeadTrackerHistoryDepth heads from the DB, returns the latest chain.
-	Load(ctx context.Context) (*evmtypes.Head, error)
+	commontypes.HeadSaver[*evmtypes.Head, common.Hash]
```
Comment (ExtremelySunnyYK) [inline core/chains/evm/headtracker/types/types.go:14]: EVM Headsaver requires its own interface due to interaction with DB

File: core/services/relay/evm/mercury/data_source_test.go [hunk 2]
```diff
@@ -177,8 +178,9 @@ func TestMercury_Observe(t *testing.T) {
 				assert.Equal(t, head.Number-1, obs.MaxFinalizedBlockNumber.Val)
 			})
 			t.Run("if current block num errored", func(t *testing.T) {
-				h2 := htmocks.NewHeadTracker(t)
-				h2.On("LatestChain").Return(nil)
+				var nilHead *evmtypes.Head
```
Comment (ExtremelySunnyYK) [inline core/services/relay/evm/mercury/data_source_test.go:181]: This workaround is needed due to generics enforcement of H where we would get panic: interface conversion: interface {} is nil, not *types.Head [recovered]

File: core/services/relay/evm/mercury/data_source_test.go [hunk 3]
```diff
@@ -297,8 +299,10 @@ func TestMercury_Observe(t *testing.T) {
 		})
 		t.Run("if head tracker returns nil, falls back to RPC method", func(t *testing.T) {
 			t.Run("if call succeeds", func(t *testing.T) {
-				h = htmocks.NewHeadTracker(t)
-				h.On("LatestChain").Return(nil)
+				var nilHead *evmtypes.Head
```
Comment (ExtremelySunnyYK) [inline core/services/relay/evm/mercury/data_source_test.go:302]: same workaround here

File: core/services/relay/evm/mercury/data_source_test.go (hunk 2 above)
Comment (jmank88) [reply core/services/relay/evm/mercury/data_source_test.go:181]: FWIW you should be able to pass `(*evmtypes.Head)(nil)` inline, if that is easier to read :shrug:

File: common/types/head_tracker.go [hunk 4]
```diff
@@ -57,12 +60,17 @@ type HeadListener[H Head[BLOCK_HASH], BLOCK_HASH Hashable] interface {
 // NewHeadHandler is a callback that handles incoming heads
 type NewHeadHandler[H Head[BLOCK_HASH], BLOCK_HASH Hashable] func(ctx context.Context, header H) error
 
+// HeadBroadcaster relays heads from the head tracker to subscribed jobs, it is less robust against
```
Comment (prashantkumar1982) [inline common/types/head_tracker.go:63]: >>less robust than head tracker, and missed heads should be expected
That seems confusing and misleading.

Just say this:

HeadBroadcaster relays new Heads to all subscribers.

File: common/types/head_tracker.go [hunk 5]
```diff
@@ -6,6 +6,9 @@ import (
 	"github.com/smartcontractkit/chainlink/v2/core/services"
 )
 
+// HeadTracker holds and stores the latest block number experienced by this particular node in a thread safe manner.
```
Comment (prashantkumar1982) [inline common/types/head_tracker.go:9]: holds and stores latest blocks.

It stores more than just the latest block.

**Review by prashantkumar1982 — COMMENTED** [review]
This is good!

File: common/types/head_tracker.go (hunk 5 above)
Comment (ExtremelySunnyYK) [reply common/types/head_tracker.go]: thanks for pointing that out prashant. Fixed

**Review by prashantkumar1982 — DISMISSED** [review]

File: core/services/relay/evm/mercury/data_source_test.go (hunk 2 above)
Comment (ExtremelySunnyYK) [reply core/services/relay/evm/mercury/data_source_test.go:181]: Thanks for the note Jordan. I think that is a cleaner approach - easier to read and shorter

**Review by prashantkumar1982 — APPROVED** [review]

**Pull request merged**